Refactor date example

tobbee opened the pull request:
* Refactored into two applications that can be run without command line options
* Fixed subscription of  announcement issue
* Further made the subscriber only continue if the announcement matches what it was configured for.
* Updated the README file with description and sequence diagram.

Fixes #194.

I will update #193 based on this.

mengelbart reviewed (changes requested):
Splitting server and client seems fine to me, but I would like to keep the option to publish and subscribe from server and client. This is a useful feature for testing interop with relays that don't publish anything themselves.

tanghaowillow commented:
How about leave the capability of both 'publish' and 'subscribe' to client and make the server simply act like a relay？Running the example can be like:
To run the server:
```shell
cd server
go run .
```
To run the client to publish/subscribe the date track to/from the server:
```shell
cd client
go run . -publish
go run . -subscribe
```
When doing interop we can start the server and publish the track by running a client as a publisher.

tobbee commented:
OK. Makes sense from the general MoQ sense. I'll redo this into a server and client where both can subscribe and/or publish and try to document that properly.

I have a pure publisher/relay/subscriber use case in mind for another project, and intend to use the publisher and subscribe variants as a basis there. Will post a link once I have a basic phase ready.

mengelbart commented:
Thanks! 

@tanghaowillow I would also like to keep the option to publish from the server so we can deploy a server to which other implementers can subscribe if they only want to test a subscriber.

tobbee commented:
I changed my mind about separation into two programs. There is very much code in common, and when both the server and the client can take publisher and subscriber roles, it is better to keep everything together.

That said, I've now made the code and documentation much clearer about the roles and possibilities.
I also fixed the announcement response so that there is no error any longer.
One could take it one step longer and subscribe to the announcement instead of the publisher just pushing them, but I leave that for a possible later enhancement.

mengelbart reviewed (approved):
Looks good, thanks @tobbee!

tobbee commented:
No problem. The commit message should be fine now.
